Writer: add marker support

timlehr opened the pull request:
### Summary

This change adds initial write support for markers. This includes all relevant marker information displayed in the Media Composer marker window:

- Name
- Comment
- User
- Creation Date
- Color

Markers are able to be round-tripped successfully and the adapter supports all colors which have an OTIO equivalent. Some of the more exotic Avid colors like Forest and Denim are remapped to more portable colors for now.

### Suggestion omitted

There was some discussion about having the marker being a truly point based event with zero duration (see https://github.com/OpenTimelineIO/otio-aaf-adapter/pull/44#issuecomment-3110210614). Since I tried to have no impact on the reader behaviour with this initial PR, we should make that change in another PR and release it separately. It's a moderate change, but unlike this PR, it would change the adapter output somewhat.

**Reference associated tests.**

- `tests/test_aaf_adapter.py`

Assisted-by: Claude:claude-opus-4-8 [code-review] [debugging]

timlehr committed:
Writer: adding marker support to AAF writer

Signed-off-by: Tim Lehr <tim.lehr@disneyanimation.com>

timlehr committed:
Writer: Fixes error with edgecode when not using embedding

Signed-off-by: Tim Lehr <tim.lehr@disneyanimation.com>

timlehr committed:
Reader: Fixed issue with markers not attaching due to Slot ID mismatch

Signed-off-by: Tim Lehr <tim.lehr@disneyanimation.com>

timlehr commented:
@markreidvfx maybe you can help me out on this one. I can't figure out why a marker on 0 would break things.

timlehr committed:
Merge branch 'main' into markerWrite_tlehr

timlehr committed:
Merge branch 'main' into markerWrite_tlehr

jminor commented:
Notes from meeting:
- Mark suggests checking to see how transitions influence the position of markers. There's some subtlety to how those interact.
- For the 0-frame issue, it would be helpful to compare to an AAF written out from Media Composer.
- Maybe a sample AAF showing the problem could be useful.

lbarral commented on src/otio_aaf_adapter/adapters/aaf_adapter/aaf_writer.py:
```diff
@@ -562,6 +590,134 @@ def aaf_transition(self, otio_transition):
+
+        aaf_sequence.components.append(marker)
+
+    def transcribe_aaf_descriptive_markers(self):
```
This method is not being called from anywhere - is that intentional?

timlehr replied:
Good catch, I think that was a cherry-pick merge issue from our internal branch!

lbarral commented on src/otio_aaf_adapter/adapters/aaf_adapter/aaf_writer.py:
```diff
@@ -562,6 +590,134 @@ def aaf_transition(self, otio_transition):
+        marker = self.aaf_file.create.DescriptiveMarker()
+        # set all possible slots described by markers
+        # FIX: Markers can appear twice in the Avid marker list at the same time
+        marker['DescribedSlots'].value = {1, 2, 3, 4, 10, 11}
```
We see the list of described slots hard-coded here.
Ideally, it would be relevant to preserve the described slot in the OTIO metadata, so it can be properly re-created back.
note: Avid's MediaComposer will not look past the first entry in this list and will, by default, create markers in the first picture track.
In case we cannot preserved the slot ID being referred to by this event slot, we would recomment to use the First Picture slot ID being generated. With it being hard-coded as is, it will always default to the Timecode track in the translation from OTIO

lbarral commented on src/otio_aaf_adapter/adapters/aaf_adapter/aaf_writer.py:
```diff
@@ -562,6 +590,134 @@ def aaf_transition(self, otio_transition):
+
+        event_mob_slot = self.aaf_file.create.EventMobSlot()
+        event_mob_slot['EditRate'].value = self.edit_rate
+        event_mob_slot['SlotID'].value = 1000
```
We shouldn't have an hard-code slot ID value here. It is also possible that there are more than one Event Mob Slots.
Media Composer will create one Event  Mob Slot for each Essence track/slot to which markers were added.

lbarral reviewed (changes requested):
We suggest some more care is taken in preserving the multi-slot nature of Marker storage in the AAF, which can't be achieved in the current implementation as it hard-codes most of it.

timlehr committed:
Writer: added support for markers

Adds initial write support for markers. This includes all relevant marker information displayed in the Media Composer marker window. Markers are able to be round-tripped successfully and the adapter supports all colors which have an OTIO equivalent.

This also already accounts for the marker color changes introduced in https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/2023, so the code is forward compatible.

Signed-off-by: Tim Lehr <tim.lehr@disneyanimation.com>
